### DISCUSSION TITLE
The 'g_pCompositor->m_vWindows' variable is unpredictable

### DISCUSSION THREAD

**Discussion by DreamMaoMao** [hyprwm/Hyprland#3648] [General]
I found a strange bug when writing the plugin, when multiple workspaces have full-screen Windows at the same time, the variable 'g_pCompositor->m_vWindows' contains only one of the full-screen window objects.

So, I can't go through all the full-screen Windows.

```cpp
for (auto& w : g_pCompositor->m_vWindows) {
    if (w->isHidden() || !w->m_bIsMapped || w->m_bFadingOut)
        continue;
   hycov_log(LOG,"dispatch_enteroverview {}",w);
```

# two fullscreen windows in two workspace:

my log show the same window object twice.but i have two different widonws.

![image](https://github.com/hyprwm/Hyprland/assets/30348075/08c90b2f-cf0c-426d-8b63-2b627f9edfcc)

# one fullscreen window in two workspace:
i can get all window normally.
![image](https://github.com/hyprwm/Hyprland/assets/30348075/1456e686-0407-48d7-8a3c-da94203bf238)

it is a bug for `g_pCompositor->m_vWindows` var?

**Comment by DreamMaoMao** [comment]
hyprland version:

Hyprland, built from branch main at commit 14a3c939ce019623126893aa4c454e62bb29779c  (hyprctl: log monitor id in workspaces request).
Tag: v0.31.0-31-g14a3c939

flags: (if any)
debug

**Comment by memchr** [comment]
What is the context and complete code? Also, check the condition.

**Comment by DreamMaoMao** [comment]
full code 
```
void dispatch_enteroverview(std::string arg) { //进入overview
    CWorkspace *PWORKSPACE;
	CWindow *PFULLWINDOW;
	for (auto& w : g_pCompositor->m_vWindows) {
        // if (w->isHidden() || !w->m_bIsMapped || w->m_bFadingOut)
        //     continue;
        if (!w->m_bIsMapped)
            continue;
		hycov_log(LOG,"dispatch_enteroverview {}",w.get());
		PWORKSPACE = g_pCompositor->getWorkspaceByID(w.get()->m_iWorkspaceID);
		if (PWORKSPACE->m_bHasFullscreenWindow) {
    	    PFULLWINDOW = g_pCompositor->getFullscreenWindowOnWorkspace(PWORKSPACE->m_iID);
    	    g_pCompositor->setWindowFullscreen(PFULLWINDOW, false, FULLSCREEN_FULL);
			PFULLWINDOW->m_bIsFullscreen = true;
    	}
	}
	g_pLayoutManager->switchToLayout("grid");
	return;
}
```

I checked the conditions. It's not the filter.

The point is that each node in the list should record a different window. Now when two workspaces have two full-screen Windows, there are two node Pointers pointing to the same window

**Reply by vaxerski** [reply]
your code / headers / something is wrong.

![image](https://github.com/hyprwm/Hyprland/assets/43317083/9465aeeb-c16a-45e9-a801-932084cbba6c)

Nothing is amiss, and a situation like yours, in which case there are two identical pointers in m_vWindows, is impossible.

**Comment by DreamMaoMao** [comment]
I gave up using this variable to traverse the window, this variable did fail.

I used another method to go through all the full screen Windows
```
void dispatch_enteroverview(std::string arg) { //进入overview
	CWindow *PFULLWINDOW;
	for (auto& w : g_pCompositor->m_vWorkspaces) {

		if (w.get()->m_bHasFullscreenWindow) {
    	    PFULLWINDOW = g_pCompositor->getFullscreenWindowOnWorkspace(w.get()->m_iID);
    	    g_pCompositor->setWindowFullscreen(PFULLWINDOW, false, FULLSCREEN_FULL);
			PFULLWINDOW->m_bIsFullscreen = true;
    	}
	}
	g_pLayoutManager->switchToLayout("grid");
	return;
}
```

**Comment by DreamMaoMao** [comment]
Okay, I'll debug it later
